File: packages/ts/test/vmtype.test.ts
import { describe, it, expect } from "vitest";
import {
  isVmType,
  vmTypeWireValue,
  vmTypeLabel,
  createRollupMsg,
  validateRollupConfig,
  presets,
  VM_TYPES,
} from "../src/index";

describe("QoreChain Native vm-type alias", () => {
  it("advertises native (not cosmwasm) in the public list", () => {
    expect(VM_TYPES).toContain("native");
    expect(VM_TYPES).not.toContain("cosmwasm");
  });

  it("accepts native, evm, svm, custom, and the cosmwasm legacy alias", () => {
    for (const v of ["native", "evm", "svm", "custom", "cosmwasm"]) {
      expect(isVmType(v)).toBe(true);
    }
    expect(isVmType("wasm2")).toBe(false);
  });

  it("maps native to the cosmwasm wire value; others pass through", () => {
    expect(vmTypeWireValue("native")).toBe("cosmwasm");
    expect(vmTypeWireValue("cosmwasm")).toBe("cosmwasm");
    expect(vmTypeWireValue("evm")).toBe("evm");
    expect(vmTypeWireValue("svm")).toBe("svm");
  });

  it("labels the Wasm runtime as QoreChain Native", () => {
    expect(vmTypeLabel("native")).toBe("QoreChain Native");
    expect(vmTypeLabel("cosmwasm")).toBe("QoreChain Native");
    expect(vmTypeLabel("evm")).toBe("EVM");
  });

  it("emits cosmwasm on the wire when the config says native", () => {
    const msg = createRollupMsg({
      creator: "qor1creator",
      rollupId: "r",
      profile: "nft",
      vmType: "native",
      stakeAmount: "1",
    });
    expect((msg.value as { vmType: string }).vmType).toBe("cosmwasm");
  });

  it("validates native and the nft preset (now native)", () => {
    const cfg = presets.nft({ rollupId: "r" }).build();
    expect(cfg.vmType).toBe("native");
    expect(validateRollupConfig(cfg).valid).toBe(true);
  });
});

File: packages/go/vmtype_test.go
package rdk

import "testing"

// TestVMTypeAdvertisedList mirrors ts/test/vmtype.test.ts: the advertised list
// carries "native", not the "cosmwasm" legacy alias.
func TestVMTypeAdvertisedList(t *testing.T) {
	hasNative := false
	hasCosmWasm := false
	for _, v := range VmTypes {
		if v == VmNative {
			hasNative = true
		}
		if v == VmCosmWasm {
			hasCosmWasm = true
		}
	}
	if !hasNative {
		t.Error("VmTypes should advertise native")
	}
	if hasCosmWasm {
		t.Error("VmTypes should not advertise the cosmwasm legacy alias")
	}
}

func TestIsVMType(t *testing.T) {
	for _, v := range []string{"native", "evm", "svm", "custom", "cosmwasm"} {
		if !IsVMType(v) {
			t.Errorf("IsVMType(%q) = false, want true", v)
		}
	}
	if IsVMType("wasm2") {
		t.Error(`IsVMType("wasm2") = true, want false`)
	}
}

func TestVMTypeWireValue(t *testing.T) {
	cases := map[string]string{
		"native":   "cosmwasm",
		"cosmwasm": "cosmwasm",
		"evm":      "evm",
		"svm":      "svm",
		"custom":   "custom",
	}
	for in, want := range cases {
		if got := VMTypeWireValue(in); got != want {
			t.Errorf("VMTypeWireValue(%q) = %q, want %q", in, got, want)
		}
	}
}

func TestVMTypeLabel(t *testing.T) {
	cases := map[string]string{
		"native":   "QoreChain Native",
		"cosmwasm": "QoreChain Native",
		"evm":      "EVM",
		"svm":      "SVM",
		"custom":   "Custom",
	}
	for in, want := range cases {
		if got := VMTypeLabel(in); got != want {
			t.Errorf("VMTypeLabel(%q) = %q, want %q", in, got, want)
		}
	}
}

// TestCreateRollupMsgWireValue: a create-rollup message built with vm_type
// "native" encodes "cosmwasm" on the wire.
func TestCreateRollupMsgWireValue(t *testing.T) {
	msg := CreateRollupMsg(CreateRollupInput{
		Creator:     "qor1creator",
		RollupID:    "r",
		Profile:     "nft",
		VmType:      "native",
		StakeAmount: 1,
	})
	if msg.VmType != "cosmwasm" {
		t.Errorf("create-rollup msg vmType = %q, want cosmwasm", msg.VmType)
	}
}

// TestNFTPresetUsesNative: the nft preset advertises native and validates.
func TestNFTPresetUsesNative(t *testing.T) {
	cfg := PresetNFT().SetRollupID("r").Get()
	if cfg.VmType != VmNative {
		t.Errorf("nft preset vmType = %q, want native", cfg.VmType)
	}
	if r := ValidateRollupConfig(cfg); !r.Valid {
		t.Errorf("nft preset should validate, got errors: %v", r.Errors)
	}
}
